src/NHibernate/Util/AsyncLock.cs

```diff
@@ -0,0 +1,47 @@
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace NHibernate.Util
+{
+	public class AsyncLock
+	{
+		private readonly AsyncSemaphore semaphore;
+		private readonly Task<Releaser> releaser;
+
+		public AsyncLock()
+		{
+			semaphore = new AsyncSemaphore(1);
+			releaser = Task.FromResult(new Releaser(this));
```

Comment:
Possible usage of a disposed object due to this. Granted, `Releaser` will support it, but that is not very usual for an `IDisposable`. If we accept this, then why not storing one unique releaser and use it for `ContinueWith` too? The already finished result would have to be renamed to something like `_completedWait`.